File: Julia/textbook_track/chapter01/ba1l.jl
# Implement PatternToNumber

const NUCLEOTIDE_ORDER = Dict{Char, Int64}('A' => 0, 'C' => 1, 'G' => 2, 'T' => 3)


pattern_to_number(pattern) = foldl((acc, nucleotide) -> 4 * acc + NUCLEOTIDE_ORDER[nucleotide], pattern; init=0)


function main()
    dna = readline()
    result = pattern_to_number(dna)
    println(result)
end


main()

File: OCaml/src/textbook_track/chapter01/ba1l.ml
let nucleotide_order = List.mapi (fun ix c -> (c, ix)) ['A'; 'C'; 'G'; 'T']


let pattern_to_number (pattern : string) : int =
    String.fold_left (fun encoding nucleotide -> 4 * encoding + List.assoc nucleotide nucleotide_order) 0 pattern


let () =
    let dna = read_line () in
    let result = pattern_to_number dna in
    print_int result; print_newline ()
